`Accordion` - add `@isStatic`, `@type`, and `@size`

alex-ju opened the pull request:
### :pushpin: Summary

`Accordion` - add the following arguments:

- [ ] `@isStatic`
  - false (default)
  - true

- [ ] `@type`
  - card (default)
  - flush

- [ ] `@size`
  - large (default)
  - medium
  - small

### :hammer_and_wrench: Detailed description

TBA

### :link: External links

Jira ticket: [HDS-3302](https://hashicorp.atlassian.net/browse/HDS-3302)
[Figma file](https://www.figma.com/design/WrZPhAVk6NVpfb5OUyTsHp/%5BAG%5D-Accordion-Design?node-id=37555-77311&t=rWuINCdPMITU2Q51-4)

***

### 👀 Component checklist

- [ ] Percy was checked for any visual regression
- [ ] A changelog entry was added via [Changesets](https://github.com/changesets/changesets) if needed (see [templates here](https://github.com/hashicorp/design-system/blob/main/wiki/Website-Changelog.md#templates-for-npm-packages))

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

didoo commented on packages/components/src/components/hds/accordion/item/index.ts:
```diff
@@ -45,6 +58,62 @@ export default class HdsAccordionItemComponent extends Component<HdsAccordionIte
+   * @default false
+   */
+  get toggleTextSize() {
+    const size = this.args.size ?? 'large';
```
should this be `medium`? also, should this come from `DEFAULT_SIZE`?

alex-ju replied:
Indeed; updated to use `DEFAULT_SIZE`

didoo commented on packages/components/src/components/hds/accordion/item/index.ts:
```diff
@@ -45,6 +58,62 @@ export default class HdsAccordionItemComponent extends Component<HdsAccordionIte
+  /**
+   * @param toggleTextSize
+   * @type {HdsTextSizes}
+   * @default false
```
```suggestion
   * @default `medium`
```

didoo commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -36,8 +56,52 @@ $hds-accordion-item-border-radius: 6px;
+    border-radius: 0;
+  }
+
+  &.hds-accordion-item--does-not-contain-interactive:not(.hds-accordion-item--is-static)>.hds-disclosure-primitive__toggle {
```
```suggestion
  &.hds-accordion-item--does-not-contain-interactive:not(.hds-accordion-item--is-static) > .hds-disclosure-primitive__toggle {
```

didoo commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -36,8 +56,52 @@ $hds-accordion-item-border-radius: 6px;
+}
+
+.hds-accordion-item__toggle-content {
+  .hds-accordion-item--is-static.hds-accordion-item--size-large & {
```
One alternative could be to render anyway the button, but with a `disabled` attribute, and hide it (visually) in CSS when `[disabled]`(not sure if it would work from an accessibility standpoint). Just an idea.

KristinLBradley replied:
+1 That would remove the need for all these extra css styles and sass variables and tie the spacing to the actual button sizes.

alex-ju replied:
I generally try to avoid rendering elements in the DOM if they are not required (especially interactive ones), but I'll give this a go if you feel that it will simplify the code.

didoo replied:
> I generally try to avoid rendering elements in the DOM if they are not required (especially interactive ones)

I 100% agree, but one of the exceptions I tend to make is this use case, when the element leaves a (visual) hole in the UI, and workaround are needed to match the layout when the element is rendered: in that case, I tend to see if it's possible to just hide the element an let the layout engine to the work for me.

alex-ju replied:
Refactored in d12ac1f

didoo commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -46,68 +110,103 @@ $hds-accordion-item-border-radius: 6px;
-      border-radius: $hds-accordion-item-border-radius;
-      content: "";
-      inset: 0;
+  .flight-icon-chevron-down {
```
In hindsights, we can use a less specific class name in this case
```suggestion
  .flight-icon {
```

didoo reviewed (commented):
Had a first pass and left a few comments.

alex-ju commented on packages/components/src/components/hds/accordion/types.ts:
```diff
@@ -0,0 +1,17 @@
+export enum HdsAccordionSizeValues {
+  Large = 'large',
+  Medium = 'medium',
+  Small = 'small',
```
Reordered them to follow the existing patter (small to large)

alex-ju commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -116,5 +215,15 @@ $hds-accordion-item-border-radius: 6px;
 
 .hds-accordion-item__content {
-  padding: 4px $hds-accordion-item-padding $hds-accordion-item-padding $hds-accordion-item-padding;
+  .hds-accordion-item--size-large & {
```
yes, CSS variables should help us clean up the code a little bit – will try and rely on those in a follow-up commit

alex-ju replied:
switched to using CSS variables in 52b6652

KristinLBradley commented on showcase/app/templates/components/accordion.hbs:
```diff
@@ -8,157 +8,122 @@
-  <Shw::Text::H3>Complex</Shw::Text::H3>
+  <Shw::Divider />
 
-  <Shw::Text::H4>With nested Accordions</Shw::Text::H4>
```
How about adding an example of a "Card" Accordion nested in a "Flush" one? (And vice versa.) I think that could be a legitimate use vase to better visually distinguish the nested vs. non-nested Accordion so would be useful to see in the Showcase.

alex-ju replied:
Yeah, we can mix different types as well. The guidance will not recommend this, but it's worth testing these edge cases.

KristinLBradley commented on packages/components/src/components/hds/accordion/item/index.ts:
```diff
@@ -45,6 +58,62 @@ export default class HdsAccordionItemComponent extends Component<HdsAccordionIte
+    const sizeMap = {
+      large: 300,
+      medium: 200,
+      small: 100,
```
Super, ultra nit: For HdsAccordionSizeValues below, sizes are ordered: small, medium, large (ie in reverse order).

alex-ju replied:
Added mixed nested examples in 8bade6c

KristinLBradley commented:
Question: Is the Accordion web documentation going to be updated in a separate PR?

andgen404 commented on packages/components/src/components/hds/accordion/index.ts:
```diff
@@ -18,6 +27,61 @@ interface HdsAccordionSignature {
+   * Sets the type of the component
+   *
+   * @param size
+   * @type {HdsAccordionSizes}
```
In Figma we updated the @type argument to `containerType`

andgen404 commented on .changeset/tiny-shrimps-flash.md:
```diff
@@ -0,0 +1,5 @@
+"@hashicorp/design-system-components": minor
+---
+
+`Accordion` - added `@size`, `@type`, and `@isStatic` arguments
```
In Figma we updated the @ type argument to containerType

alex-ju replied:
Thanks for the heads-up – I saw your comment in the CRD yesterday. In the Ember version of this component, we don't have the concept of a container (it could be interpreted as the container that holds the revealed content or something else). Can you provide some background on what prompted this change (in the CRD or here)?

andgen404 replied:
@alex-ju @majedelass I want to update this comment as resolved via Slack thread [here](https://hashicorp.slack.com/archives/C025N5V4PFZ/p1718715573875349) 
---
We've decided to move from containerType in Figma and align it with the code. So now both Code and Figma are Type.

alex-ju commented:
> Question: Is the Accordion web documentation going to be updated in a separate PR?

Yes, we'll have a separate PR for the documentation updates.

alex-ju commented:
@didoo, @KristinLBradley I hope I addressed all the feedback, so please have another look when you have time.

Note that `@type` may be renamed to `@containerType` in a follow-up commit, as mentioned by Andrew's in the comment above.

didoo reviewed (commented):
Left another bunch of comments. We're almost there :)

alex-ju commented on showcase/tests/integration/components/hds/accordion/index-test.js:
```diff
@@ -143,4 +202,22 @@ module('Integration | Component | hds/accordion/index', function (hooks) {
+    assert.dom('.hds-accordion-item--is-static').exists();
+    assert
+      .dom('.hds-accordion-item__button')
+      .hasStyle({ visibility: 'hidden' });
```
I tried, but it doesn't 😞

didoo replied:
Oh dear! Interesting. Never mind then 🤷‍♂️

alex-ju commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -10,24 +10,50 @@
 $hds-accordion-item-border-radius: 6px;
-
-// ACCORDION COMPONENT (wrapper)
+$hds-accordion-item-gap: 12px;
```
At this point probably no. Will remove it

alex-ju commented on packages/components/src/components/hds/accordion/item/index.hbs:
```diff
@@ -11,18 +11,19 @@
         @onClickToggle={{t.onClickToggle}}
         @contentId={{this.contentId}}
         @ariaLabel={{this.ariaLabel}}
+        @size={{this.size}}
```
with `visibility: hidden;` we remove it from the a11y tree, so no need to disable it

didoo replied:
Ohhhh! TIL! Thanks

alex-ju commented on showcase/tests/acceptance/components/hds/accordion-test.js:
```diff
@@ -12,9 +12,18 @@ module('Acceptance | Component | hds/accordion', function (hooks) {
   setupApplicationTest(hooks);
 
   test('Components/hds/accordion page passes automated a11y checks', async function (assert) {
+    let axeOptions = {
```
updated here, but we have it as `let` in all other places

alex-ju commented on packages/components/src/components/hds/accordion/item/index.ts:
```diff
@@ -45,6 +58,62 @@ export default class HdsAccordionItemComponent extends Component<HdsAccordionIte
+   */
+  get toggleTextSize() {
+    const size = this.args.size ?? DEFAULT_SIZE;
+    const sizeMap = {
```
moved this in 1316e225b

alex-ju commented on showcase/tests/integration/components/hds/accordion/index-test.js:
```diff
@@ -48,6 +48,66 @@ module('Integration | Component | hds/accordion/index', function (hooks) {
+      .dom('#test-accordion .hds-accordion-item')
+      .hasClass('hds-accordion-item--size-large');
+  });
+
```
your pseudocode was working fine – added to both `@size` and `@type` in 2c19eef48

alex-ju commented:
@didoo, @KristinLBradley I hope I addressed the latest round of feedback, so please have another look when you have time.

didoo reviewed (approved):
Left a comment for @majedelass, apart from that: all good 👍

KristinLBradley reviewed (approved):
Looking good! I am curious about what @majedelass thinks in response to [Cristiano's question about the focus ring](https://github.com/hashicorp/design-system/pull/2156/files#r1647494332) though.

majedelass commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -36,75 +59,104 @@ $hds-accordion-item-border-radius: 6px;
+
+  // we need this specificity to override the focus-ring mixin
+  .hds-accordion-item__button.hds-accordion-item__button--parent-does-not-contain-interactive::before {
+    border-radius: 0;
```
I think the focus ring should be inset because our components behave this way (I'm immediately thinking about our listItems). If we were to inset this, I see a couple pixel gap top/bottom in the itemToggle. Can we reduce that gap? Otherwise I'm ok with this approach 👍

andgen404 replied:
**RE**: Update to this comment. I made these changes in the Figma component to reflect Cristiano's suggestion. It can be found in the Figma file [here](https://www.figma.com/design/noyY6dUMDYjmySpHcMjhkN/branch/WrZPhAVk6NVpfb5OUyTsHp/HDS-Components?m=auto&node-id=37555-77311&t=g7ZfY27G1y7hXqtO-1)

alex-ju replied:
Thanks for sharing the Figma reference @andgen404! Just to make sure I understand correctly, we're planning to have the focus 'inset' for the flush variant, and 'outset' for the card variant, is that right?

heatherlarsen commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -36,75 +59,104 @@ $hds-accordion-item-border-radius: 6px;
-    $hds-accordion-item-padding
-    $hds-accordion-item-padding
-    12px; // by design
+  padding: var(--hds-accordion-item-padding) var(--hds-accordion-item-padding) var(--hds-accordion-item-padding) 12px;
```
@majedelass @andgen404, please double-check that all variants should have 12px left padding.

alex-ju replied:
Will do. I was told about these changes yesterday and didn't get the chance to operate them. We also discussed some adjustments for the nested flush accordions that I will need to apply in a follow-up commit.

alex-ju replied:
Addressed in 3e847d8 – @majedelass @andgen404 please check

heatherlarsen commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -116,5 +168,7 @@ $hds-accordion-item-border-radius: 6px;
 .hds-accordion-item__content {
-  padding: 4px $hds-accordion-item-padding $hds-accordion-item-padding $hds-accordion-item-padding;
+  .hds-accordion-item & {
+    padding: 4px var(--hds-accordion-item-padding) var(--hds-accordion-item-padding) var(--hds-accordion-item-padding);
```
@majedelass @andgen404 I think the `small` variant is set in Figma to 12px 4px 8px 8px. Can you double-check this, please?

alex-ju replied:
Addressed in 3e847d8

andgen404 commented on packages/components/src/styles/components/accordion.scss:
```diff
@@ -36,75 +78,100 @@ $hds-accordion-item-border-radius: 6px;
-    &.mock-focus {
-      @include hds-button-state-focus();
-    }
+// entire toggle area is interactive
```
@alex-ju Out of curiosity, are you utilizing the ::before and ::after pseudo-elements to style the focus state of the accordion component?

alex-ju replied:
yes, we are using the `::before` pseudo-element for the focus states (in many of our components)

![Screenshot 2024-06-25 at 10 16 19](https://github.com/hashicorp/design-system/assets/788096/b72803c9-d877-4de7-ab27-1b85a004783c)

andgen404 replied:
Ah, okay, thank you! I appreciate the example image as well.

alex-ju commented:
Rebased after #2181.

andgen404 reviewed (approved):
Visually Lgtm 🎉

alex-ju commented:
> Visually Lgtm 🎉

@andgen404 I updated the `Accordion::Item` flush variant to only have a border at the bottom, even when in groups. [Have a look](https://hds-showcase-38tyfc0qw-hashicorp.vercel.app/components/accordion) when you have the time and let me know if it aligns with your view.

zamoore reviewed (approved):
Nice!

didoo commented on packages/components/src/components/hds/accordion/item/button.hbs:
```diff
@@ -12,5 +12,5 @@
   ...attributes
 >
-  <FlightIcon @name="chevron-down" @size="24" @isInlineBlock={{false}} />
+  <FlightIcon @name="chevron-down" @size={{if (eq @size "large") "24" "16"}} @isInlineBlock={{false}} />
```
Good point! Technically they should come from here: https://github.com/hashicorp/design-system/blob/main/packages/ember-flight-icons/src/components/flight-icon.ts#L17

andgen404 commented:
> > Visually Lgtm 🎉
> 
> @andgen404 I updated the `Accordion::Item` flush variant to only have a border at the bottom, even when in groups. [Have a look](https://hds-showcase-38tyfc0qw-hashicorp.vercel.app/components/accordion) when you have the time and let me know if it aligns with your view.

Thank you for updating that! Looks great 👍
